tools/keepalived/keepalived/libipvs-2.6/libipvs.c

```diff
@@ -393,40 +441,45 @@ int ipvs_stop_daemon(ipvs_daemon_t *dm)
 struct ip_vs_get_services *ipvs_get_services(void)
 {
 	struct ip_vs_get_services *get;
-	struct ip_vs_get_services_kern *getk,*getk_rcv;
+	struct dp_vs_get_services *dpvs_get, *dpvs_get_rcv;
+	struct dp_vs_service_entry *dpvs_entry;
+	struct ip_vs_service_entry *ipvs_entry;
 	size_t len, len_rcv;
 	int i;
 
-
-	len = sizeof(*get) +
+	len = sizeof(struct ip_vs_get_services) +
 		sizeof(ipvs_service_entry_t) * ipvs_info.num_services;
 	if (!(get = calloc(len, 1)))
 		return NULL;
-	len = sizeof(*getk) +
-		sizeof(struct ip_vs_service_entry_kern) * ipvs_info.num_services;
-	if (!(getk = malloc(len))) {
+
+	len = sizeof(struct dp_vs_get_services) + 
+		sizeof(struct dp_vs_service_entry) * ipvs_info.num_services;
+	if (!(dpvs_get = calloc(len, 1))) {
 		free(get);
 		return NULL;
 	}
-
-	ipvs_func = ipvs_get_services;
-	getk->num_services = ipvs_info.num_services;
+	dpvs_get->num_services = ipvs_info.num_services;
 	len_rcv = len;
-	if (dpvs_getsockopt(DPVS_SO_GET_SERVICES, getk, len, (void **)&getk_rcv, &len_rcv) < 0) {
+	
+	if (dpvs_getsockopt(DPVS_SO_GET_SERVICES, dpvs_get, len, (void **)&dpvs_get_rcv, &len_rcv) < 0) {
 		free(get);
-		free(getk);
-        //dpvs_sockopt_msg_free(getk_rcv);
+		free(dpvs_get);
 		return NULL;
 	}
-	memcpy(get, getk_rcv, sizeof(struct ip_vs_get_services));
-	for (i = 0; i < getk_rcv->num_services; i++) {
-		memcpy(&get->entrytable[i], &getk_rcv->entrytable[i],
-		       sizeof(struct ip_vs_service_entry_kern));
-		get->entrytable[i].af = AF_INET;
-		get->entrytable[i].addr.ip = get->entrytable[i].__addr_v4;
+
+	get->num_services = ipvs_info.num_services;
+	for (i = 0; i < dpvs_get_rcv->num_services; i++) {
+		ipvs_entry = &get->entrytable[i];
+		dpvs_entry = &dpvs_get_rcv->entrytable[i];
+		DPVS_2_IPVS(ipvs_entry, dpvs_entry);
+		if (dpvs_get_rcv->entrytable[i].af == AF_INET) {
+			get->entrytable[i].__addr_v4 = get->entrytable[i].addr.ip;
+			get->entrytable[i].pe_name[0] = '\0';
```

Comment:
What's the purpose of the two 'fields'? IPv6 not support them?

Reply:
Original ipvsadm/keepalived struct support. ipv4 show always show __addr_v4; If we don't set these filed,  we have to change many structs of ipvsadm/keepalived.